src/backend/distributed/commands/schema_based_sharding.c

```diff
@@ -227,18 +575,230 @@ citus_internal_unregister_tenant_schema_globally(PG_FUNCTION_ARGS)
 	if (HeapTupleIsValid(namespaceTuple))
 	{
 		ReleaseSysCache(namespaceTuple);
-
 		ereport(ERROR, (errmsg("schema is expected to be already dropped "
 							   "because this function is only expected to "
 							   "be called from Citus drop hook")));
 	}
-
 	uint32 tenantSchemaColocationId = SchemaIdGetTenantColocationId(schemaId);
 
 	DeleteTenantSchemaLocally(schemaId);
 	SendCommandToWorkersWithMetadata(TenantSchemaDeleteCommand(schemaNameStr));
 
 	DeleteColocationGroup(tenantSchemaColocationId);
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * citus_schema_distribute gets a regular schema name, then converts it to a tenant
+ * schema.
+ */
+Datum
+citus_schema_distribute(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	Oid schemaId = PG_GETARG_OID(0);
+
+	/* Prevent concurrent schema deletion and table creation under the schema */
+	LockDatabaseObject(NamespaceRelationId, schemaId, 0, AccessExclusiveLock);
+
+	/*
+	 * It is possible that by the time we acquire the lock on schema,
+	 * concurrent DDL has removed it. We can test this by checking the
+	 * existence of schema.
+	 */
+	if (!SearchSysCacheExists1(NAMESPACEOID, ObjectIdGetDatum(schemaId)))
+	{
+		ereport(ERROR, (errcode(ERRCODE_UNDEFINED_SCHEMA),
+						errmsg("concurrent operation removed the schema")));
+	}
+
+	/* Ensure schema checks */
+	EnsureTenantSchemaAllowed(schemaId);
+	EnsureSchemaOwner(schemaId);
+
+	/* Return if the schema is already a tenant schema */
+	char *schemaName = get_namespace_name(schemaId);
+	if (IsTenantSchema(schemaId))
+	{
+		ereport(NOTICE, (errmsg("schema %s is already a tenant schema", schemaName)));
+		PG_RETURN_VOID();
+	}
+	else
+	{
+		ereport(NOTICE, (errmsg("converting the schema %s to a tenant schema",
+								schemaName)));
+	}
+
+	/* Collect all tables by their types in the schema */
+	SchemaTables *schemaTables = TablesUnderSchema(schemaId);
```

Comment:
Should we rename EnsureTenantSchemaAllowed to EnsureSchemaCanBeConvertedToTenantSchema / EnsureSchemaCanBeUsedAsenantSchema or such and append the following checks to it, to simplify this UDF?

```c
.. // preserve the existing checks

	/* Collect all tables by their types in the schema */
	SchemaTables *schemaTables = TablesUnderSchema(schemaId);

	/* Allow only Citus local and PG local tables in the schema */
	List *disallowedCitusTableTypes = DisallowedTableTypesForDistributeSchema();
	if (SchemaHasCitusTableType(schemaTables, disallowedCitusTableTypes))
	{
		ereport(ERROR, (errmsg("schema already has distributed tables"),
						errhint("Undistribute distributed tables under "
								"the schema before distributing the schema.")));
	}

	/* We check that all table relation kinds are supported by a tenant schema */
	List *postgresTableIdList = schemaTables->postgresTableIdList;
	postgresTableIdList = list_concat(postgresTableIdList,
									  schemaTables->citusLocalTableIdList);

	Oid relationId = InvalidOid;
	foreach_oid(relationId, postgresTableIdList)
	{
        // comment why we skip partition rels as before, due to oid change..
		if (PartitionTable(relationId))
		{
              EnsureRelationExists(relationId);
		}
        
     	EnsureTableKindsSupportedForTenantSchema(relationId);
	}
```

Reply:
(it's ok to call TablesUnderSchema twice, once in EnsureTenantSchemaAllowed and in citus_schema_distribute to convert the tables into single shard tables)

Reply:
Should we also check the unsupported foreign keys in this function, to avoid throwing an error for an unsupported foreign key type after converting N tables that required moving M GB of data?

Checking the following would be sufficient --we don't need to re-invent the detailed fkey checks for this function:
// allow foreign keys "from/to" tables in the same tenant schema
// allow the foreign keys "to" reference tables created in non-tenant schemas
// throw an error for other cases

Reply:
(self-note: not forget reviewing those changes)

Reply:
```sql
CREATE SCHEMA s2;
CREATE SCHEMA s3;

CREATE TABLE s2.t1 (a int PRIMARY KEY, b text);
CREATE TABLE s2.t2 (a int PRIMARY KEY, b text);
CREATE TABLE s2.t3 (a int, b text);

CREATE TABLE s3.t1 (a int PRIMARY KEY, b text);
CREATE TABLE s3.t2 (a int PRIMARY KEY, b text);
CREATE TABLE s3.t3 (a int, b text);

ALTER TABLE s3.t3 ADD FOREIGN KEY (a) REFERENCES s2.t1(a);

-- rn this doesn't fail because deep fkey checks allow fkeys from local tables (s3.t3) to distributed tables, assuming that user will soon distribute the local one.
-- early fkey checks would help prevent such cases
SELECT citus_schema_distribute('s2');
```